### PR TITLE
[feat] add prometheus support for proxy & httpserver.

### PR DESCRIPTION
**Pull request by LokiWager**
## Background

Add Prometheus support for Easegress Objects & Filters, for details, see `metrics.md`.

## Changing

1. Add Prometheus helper function.
    * check metric & labels
    * check duplication for metric
    * create Prometheus metric and registered to `DefaultRegisterer` automatically.
2. Create metrics for HttpServer
    * collect these metrics in `pool.collectMetrics` function.
    

| Metric | Type | Description | Labels                                                                  |
|--------| ---- | ----------- |-------------------------------------------------------------------------|
| httpserver_health | gauge | show the status for the http server: 1 for ready, 0 for down | clusterName, clusterRole, instanceName, name, kind                      |
| httpserver_total_requests | counter | the total count of http requests | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |
| httpserver_total_response | counter | the total count of http resposne | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |
| httpserver_total_error_requests | counter | the total count of http error requests | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |
| httpserver_requests_duration | histogram | request processing duration histogram | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |
| httpserver_requests_size_bytes | histogram | a histogram of the total size of the request. Includes body | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |
| httpserver_response_size_bytes | histogram | a histogram of the total size of the returned response body | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |

3. Create metrics for Proxy
    * collect these metrics in `serveHTTP` function

| Metric | Type | Description | Labels                                                                              |
|--------| ---- | ----------- |-------------------------------------------------------------------------------------|
| proxy_total_connections| counter | the total count of proxy connections | clusterName, clusterRole, instanceName, name, kind, loadBalancePolicy, filterPolicy |
| proxy_total_error_connections | counter | the total count of proxy error connections | clusterName, clusterRole, instanceName, name, kind, loadBalancePolicy, filterPolicy |
| proxy_request_body_size | histogram | a histogram of the total size of the request | clusterName, clusterRole, instanceName, name, kind, loadBalancePolicy, filterPolicy |
| proxy_response_body_size | histogram | a histogram of the total size of the response | clusterName, clusterRole, instanceName, name, kind, loadBalancePolicy, filterPolicy |

4. Add exporter URI

```
Get /apis/v2/metrics
```

## implementation

Why not take the easemonitor approach to implementation
1. not enough timeliness to collect every 5 seconds
2. Can't take full advantage of the Prometheus ecosystem
3. labels acquisition is more invasive to the original code, and the customization is serious

### CODE REVIEW

File: doc/reference/metrics.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+| httpserver_health | gauge | show the status for the http server: 1 for ready, 0 for down | clusterName, clusterRole, instanceName, name, kind                      |
+| httpserver_total_requests | counter | the total count of http requests | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |
+| httpserver_total_response | counter | the total count of http resposne | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |
+| httpserver_total_error_requests | counter | the total count of http error requests | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |
+| httpserver_requests_duration | histogram | request processing duration histogram | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |
+| httpserver_requests_size_bytes | histogram | a histogram of the total size of the request. Includes body | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |
+| httpserver_response_size_bytes | histogram | a histogram of the total size of the returned response body | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |
```
Comment (localvar) [inline doc/reference/metrics.md:34]:
```suggestion
| httpserver_responses_size_bytes | histogram | a histogram of the total size of the returned response body | clusterName, clusterRole, instanceName, name, kind, routerKind, backend |
```

and the `description` is different from `httpserver_requests_size_bytes`, is it correct?

File: pkg/filters/proxy/metrics.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	metrics struct {
+		TotalConnections      *prometheus.CounterVec
+		TotalErrorConnections *prometheus.CounterVec
+		RequestBodySize       prometheus.ObserverVec
+		ResponseBodySize      prometheus.ObserverVec
+	}
+)
+
+// newMetrics create the ProxyMetrics.
+func (p *Proxy) newMetrics(name string) *metrics {
```
Comment (localvar) [inline pkg/filters/proxy/metrics.go:37]: propose to put all the content of this file into pool.go, and change the receiver of this function to `ServerPool`.

File: pkg/object/httpserver/mux.go [hunk 3]
```diff
@@ -47,6 +47,7 @@ type (
 	mux struct {
 		httpStat *httpstat.HTTPStat
 		topN     *httpstat.TopN
+		metrics  *metrics
```
Comment (localvar) [inline pkg/object/httpserver/mux.go:50]: I haven't found any usage of this field, if this is true, please remove it.

File: pkg/object/httpserver/metrics.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+		TotalResponse      *prometheus.CounterVec
+		TotalErrorRequests *prometheus.CounterVec
+		RequestsDuration   prometheus.ObserverVec
+		RequestSizeBytes   prometheus.ObserverVec
+		ResponseSizeBytes  prometheus.ObserverVec
+	}
+)
+
+// newMetrics create the HttpServerMetrics.
+func (r *runtime) newMetrics(name string) *metrics {
```
Comment (localvar) [inline pkg/object/httpserver/metrics.go:39]: please put the content of this file to `runtime.go` and `mux.go` and remove this file.

File: pkg/util/prometheushelper/helper.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+)
+
+var (
+	validMetric = regexp.MustCompile(`^[a-zA-Z_:][a-zA-Z0-9_:]*$`)
+	validLabel  = regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_]*$`)
+)
+
+// NewCounter create the counter metric
+func NewCounter(metric string, help string, labels []string) *prometheus.
+	CounterVec {
```
Comment (localvar) [inline pkg/util/prometheushelper/helper.go:44]: per https://github.com/golang/go/wiki/CodeReviewComments#line-length, better not wrap this line, and there are other cases.
```suggestion
func NewCounter(metric string, help string, labels []string) *prometheus. CounterVec {
```

File: pkg/util/prometheushelper/helper.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+			Help: help,
+		},
+		labels,
+	)
+	prometheus.MustRegister(summaryMap[metricName])
+	logger.Infof("[%s] Summary <%s> already created!", module, metricName)
+	return summaryMap[metricName]
+}
+
+func getAndValid(metricName string, labels []string) (string, error) {
```
Comment (localvar) [inline pkg/util/prometheushelper/helper.go:147]: `valid` is an adjective.

```suggestion
func getAndValidate(metricName string, labels []string) (string, error) {
```

File: pkg/util/prometheushelper/helper.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+	for _, l := range labels {
+		if ValidLabelName(l) == false {
+			return "", fmt.Errorf("Invalid label name: %s", l)
+		}
+	}
+	return metricName, nil
+}
+
+// ValidMetricName check if the metric name is valid
+func ValidMetricName(name string) bool {
```
Comment (localvar) [inline pkg/util/prometheushelper/helper.go:161]: Ditto

File: pkg/util/prometheushelper/helper.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+	return validMetric.MatchString(name)
+}
+
+// ValidLabelName check if the label name is valid
+func ValidLabelName(label string) bool {
```
Comment (localvar) [inline pkg/util/prometheushelper/helper.go:166]: Ditto

File: pkg/util/prometheushelper/helper.go [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+	prometheus.MustRegister(counterMap[metricName])
+
+	logger.Infof("[%s] Counter <%s> is created!", module, metricName)
+	return counterMap[metricName]
+}
+
+// NewGauge create the gauge metric
+func NewGauge(metric string, help string, labels []string) *prometheus.
+	GaugeVec {
+
```
Comment (localvar) [inline pkg/util/prometheushelper/helper.go:73]:
```suggestion
```

File: pkg/util/prometheushelper/helper.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+func NewCounter(metric string, help string, labels []string) *prometheus.
+	CounterVec {
+
+	metricName, err := getAndValid(metric, labels)
+	if err != nil {
+		logger.Errorf("[%s] %v", module, err)
+		return nil
+	}
+
+	if m, find := counterMap[metricName]; find {
```
Comment (localvar) [inline pkg/util/prometheushelper/helper.go:52]: will `NewCounter` be called concurrently? if yes, here could be a race condition.

**Review by localvar — APPROVED** [review]

**Comment by zhao-kun** [comment]
@suchen-sci please review the PR ASAP

**Review by suchen-sci — APPROVED** [review]

**Pull request merged**